cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuthorizationResult.java

```diff
@@ -0,0 +1,141 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.spi.authorization;
+
+import java.util.Set;
+import javax.annotation.Nullable;
+
+public class AuthorizationResult {
+
+  public enum AuthorizationStatus {
+    AUTHORIZED,
+    UNAUTHORIZED,
+    NOT_REQUIRED
+  }
+
+  private final AuditLogContext auditLogContext;
+  private final AuthorizationStatus authorized;
+  // Fields to handle Unauthorized exception
+
+  private final String principal;
+  private final Set<String> missingPermissions;
+  private final String entity;
+  private final Throwable ex;
+  private final boolean requiresAllPermissions;
+  private final boolean includePrincipal;
+  private final String addendum;
+
+  private AuthorizationResult(Builder builder) {
+    this.auditLogContext = builder.auditLogContext;
+    this.authorized = builder.authorized;
+    this.principal = builder.principal;
+    this.missingPermissions = builder.missingPermissions;
+    this.entity = builder.entity;
+    this.ex = builder.ex;
+    this.requiresAllPermissions = builder.requiresAllPermissions;
+    this.includePrincipal = builder.includePrincipal;
+    this.addendum = builder.addendum;
+  }
+
+  /**
+   * @return the audit log context object from the enforcement response
+   */
+  public AuditLogContext getAuditLogContext() {
+    return auditLogContext;
+  }
+
+  /**
+   * @return if the enforcement/ authorization was allowed
+   */
+  public AuthorizationStatus isAuthorized() {
+    return authorized;
+  }
+
+  public UnauthorizedException createUnAuthorizedException() {
+    return new UnauthorizedException(principal, missingPermissions, entity, ex, requiresAllPermissions,
+                                     includePrincipal, addendum);
+  }
+
+  public static class Builder {
+    private AuditLogContext auditLogContext;
+    private AuthorizationStatus authorized;
+
+    private String principal;
+    private Set<String> missingPermissions;
+    private String entity;
+    private Throwable ex;
+    private boolean requiresAllPermissions;
+    private boolean includePrincipal;
+    private String addendum;
+
+    public Builder setAuditLogContext(AuditLogContext auditLogContext) {
+      this.auditLogContext = auditLogContext;
+      return this;
+    }
+
+    public Builder setAuthorized(AuthorizationStatus authorized) {
+      this.authorized = authorized;
+      return this;
+    }
+
+    public Builder setForUnauthExpPrincipal(String principal) {
+      this.principal = principal;
+      return this;
+    }
+
+    public Builder setForUnauthExpMissingPermissions(Set<String> missingPermissions) {
+      this.missingPermissions = missingPermissions;
+      return this;
+    }
+
+    public Builder setForUnauthExpEntity(String entity) {
+      this.entity = entity;
+      return this;
+    }
+
+    public Builder setForUnauthExpThrowable(Throwable ex) {
+      this.ex = ex;
```

Comment:
What is this one? Is it UnauthorizedException root cause? I don't think we need it here. Eventually everything will be AuthorizationResult, so there going to be no root causes

Reply:
Well, the question is if we want the latest SPI to be able to set the root cause. Note that it will be UnauthorizedException, so usually we should not have a root cause as in some IO Exceptions. Currently we don't use it in the GCP extension.
Also generally you should not pass an exception in the field as it's much harder to serialize.
Overall I'd just skip it as it does not add value, but complicates things

Reply:
Note that we should also review other fields, especially principal and entity. Those are usually known from the context and can be filled from it instead of asking SPI to pass it in the AuthorizationResult

Reply:
But removing other fields is up to us, we can see how it's used in extensions right now

Reply:
Looking at the usage of all this fields in Unauthorized exception, they are mostly used to construct it's message. We can see if this can be simplified. @dli357 WDYT?